tests/util.py

```diff
@@ -31,6 +31,27 @@ def test_split_tokens():
     assert split_tokens("#test") == ['#', 'test']
 
 
+@test
+def test_validate_expression():
+    assert validate_expression('') is True
+    assert validate_expression('1') is True
+    assert validate_expression('1+1') is True
+    assert validate_expression('(1+1)') is True
+    assert validate_expression('(1+1)*2') is True
+    assert validate_expression('(1 + 1)') is True
+    assert validate_expression('10 % 2') is True
+    assert validate_expression('0x100 << 2') is True
+    assert validate_expression('0x100~1') is True
+    assert validate_expression('0xabcdef') is True
+    assert validate_expression('!1^2*3+4/5&6|7') is True
+    assert validate_expression('(((((1+1) * 2') is True  # valid characters, even if expression is not valid
+
+    assert validate_expression(':') is False
+    assert validate_expression('_') is False
+    assert validate_expression('=') is False
+    assert validate_expression('.') is False
+
```

Comment:
assert validate_expression('evil()') is False

Reply:
assert validate_expression('dead()') is False

Reply:
Good catch! This was a bit harder to test for than the simple check if a character is in a set of allowed ones, but I now check that characters a-f can only exist in a sequence of valid hex digits that started with 0x. See commit bc6b848

Reply:
Btw, I also added tests for `eval_arg` to test that all expressions (operators) we support actually get evaluated correctly. Additionally I have added to the compat tests (expr.S) all supported cases, to ensure that binutils gets to the same result.

(Surprisingly the ^ operator does not result in the same output. My testing so far shows python to get to the right result, while binutils gets to either 0 or 1 depending on what I put in. With all reading so far, I could not figure out why: it should really be treated like a bitwise XOR by binutils, but for some reason isn't. I also grepped the relevant include files from ESP-IDF and they never use ^ in defining constants, so this operator is not so important. Consequently I have removed support for that operator for now. Using it will for now throw the "Unsupported operator" exception)